url58b12a0c74_tatianaslod_daooTP_tgz-pJ8-tp1_MainTp1J8/src/server/PoolThreadTaskExecutor.java

```diff
@@ -0,0 +1,80 @@
+package server;
+
+import com.sun.istack.internal.NotNull;
+import daoo.ioc.Task;
+import daoo.ioc.TaskExecutor;
+
+import java.util.LinkedList;
+
+import static java.util.concurrent.Executors.newFixedThreadPool;
+
+public class PoolThreadTaskExecutor implements TaskExecutor {
+    private final WorkQueue workqueue;
+
+    public PoolThreadTaskExecutor(int n) {
+        workqueue = new WorkQueue(n);
+    }
+
+    @Override
+    public void execute(@NotNull Task task) {
+//        ExecutorService executorService = newFixedThreadPool(16);
+//        executorService.execute(task);
+//        threads.get(executedAmount);
+//        ThreadFactory threadFactory = defaultThreadFactory();
+        workqueue.execute(task);
+    }
+
+    public class WorkQueue {
+        private final int nThreads;
+        @NotNull
+        private final PoolWorker[] threads;
+        @NotNull
+        private final LinkedList<Runnable> queue;
+
+        public WorkQueue(int nThreads) {
+            this.nThreads = nThreads;
+            queue = new LinkedList<Runnable>();
+            threads = new PoolWorker[nThreads];
+
+            for (int i = 0; i < nThreads; i++) {
+                threads[i] = new PoolWorker();
+                threads[i].start();
+            }
+        }
+
+        public void execute(@NotNull Runnable r) {
+            synchronized (queue) {
+                queue.addLast(r);
+                queue.notify();
+            }
+        }
+
+        private class PoolWorker extends Thread {
+            public void run() {
+                Runnable r;
+
+                while (true) {
+                    synchronized (queue) {
+                        while (queue.isEmpty()) {
+                            try {
+                                queue.wait();
+                            } catch (InterruptedException ignored) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)